### PR TITLE
Add IWYU to CI

### LINKED ISSUE #581: Compile times are growing significantly
## Feature request

As anyone who has built libgdf recently has surely noticed, the time to compile the library from scratch has grown significantly in the last few months. For example, compiling on all 10 cores of my i9-7900X  @ 3.30GHz takes 11 minutes as reported by `time make -j`.

Compiling with `-ftime-report` may be a good place to start to see where all the compilation time is being spent.

This is likely due to the large amount of template instantiation that is required for instantiating functions for all possible types. We should make sure that best practices are being followed in template instantiation such that a template for a given type is only having to be instantiated once via explicit instantiation.

Much of our code is implemented in headers, which causes it to be recompiled everywhere that header is included. Using pre-compiled headers may help: 
https://gcc.gnu.org/onlinedocs/gcc/Precompiled-Headers.html
http://itscompiling.eu/2017/01/12/precompiled-headers-cpp-compilation/

Furthermore, code should be scrubbed from excessive and unnecessary `#include`s. Compiling with `-MD` will show what files are being included 

Here's a Clang tool that ensures you only include the necessary headers: https://github.com/include-what-you-use/include-what-you-use

Here's a Clang tool to profile time spent in template instantiation: https://github.com/mikael-s-persson/templight

### PR DESCRIPTION
**Pull request by vyasr**
## Description

This PR adds [`include-what-you-use`](https://github.com/include-what-you-use/include-what-you-use/) to the CI job running clang-tidy. Like clang-tidy, IWYU runs via CMake integration and only runs on cpp files, not cu files. This should help us shrink binaries and reduce compilation times in cases where headers are being included unnecessarily, and it helps keep our include lists clean. The IWYU suggestions for additions are quite noisy and the team determined this to be unnecessary, so this PR instead post-filters the outputs to only show the removals. The final suggestions are uploaded to a file that is uploaded to the GHA page so that it can be downloaded, inspected, and easily applied locally.

Resolves #581.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: ci/clang_tidy.sh
Comment (vyasr) [inline ci/clang_tidy.sh:45]:
```suggestion
python cpp/scripts/parse_iwyu_output.py build_output.txt iwyu_output.txt
```

File: cpp/CMakeLists.txt
Comment (vyasr) [inline cpp/CMakeLists.txt:185]: Do we need these options, or should we just always enable all linters?

File: cpp/tests/CMakeLists.txt
Comment (vyasr) [inline cpp/tests/CMakeLists.txt:86]: This change disabled clang-tidy on tests.

File: cpp/.clang-tidy
Comment (vyasr) [inline cpp/.clang-tidy:42]: This change makes clang-tidy warnings not be errors so the CI build doesn't get marked as a failure and always runs to completion.

File: cpp/.clang-tidy
Comment (vyasr) [inline cpp/.clang-tidy:43]: This change prevents linting of any headers in the tests directory.

File: ci/cpp_linters.sh
Comment (vyasr) [inline ci/cpp_linters.sh:38]: This build is pretty trivial, so now that we have it working if this PR gets approved before a release I would be fine with merging and making a follow-up to use the conda package once it is released rather than holding up the PR waiting for a release.

**Comment by vyasr** [comment]
In addition to the inline notes above, there are a couple other points that I anticipate coming up from reviewers:
- Ideally we should probably rewrite the parsing script so that it can be run live on the CI logs (i.e. pipe `cmake` output directly to the script) and then it would print all unrelated lines back out to stdout so that we would still get live logging but with IWYU output removed. Then, at the end we could just print out (in addition to uploading) the final compiled IWYU suggestions with all the undesired lines removed. That will take some reworking of the script, though, which currently relies on reading the output all at once. I didn't think it was worth holding that up right now, but if reviewers think it is worthwhile we can work on getting that change into this PR.
- IWYU does support [pragmas](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUPragmas.md) that we might want to add so that it stops suggesting any repeated changes. 

I'll leave it to reviewers to determine what to gate this PR on vs. doing in follow-ups.

File: ci/cpp_linters.sh
Comment (harrism) [inline ci/cpp_linters.sh:48]: suggestion: Maybe improve the comment so that readers don't have to visually parse the regex. Where does the "__w" and the repeated "cudf" come from?

File: cpp/.clang-tidy
Comment (harrism) [reply cpp/.clang-tidy:43]: Would it be clearer to use the `ExcludeHeaderFilterRegex` to *exclude* the `tests` directory?

**Review by harrism — APPROVED** [review]
Looks great. Couple of comments.

File: ci/cpp_linters.sh
Comment (vyasr) [reply ci/cpp_linters.sh:48]: Good idea, I'll add a note about this (it's from the paths in CI due to GHA defaults).

File: cpp/.clang-tidy
Comment (vyasr) [reply cpp/.clang-tidy:43]: No because there are many more top-level directories in the cpp dir, and because the way that the filtering works the exclusion is layered on top of the inclusion so we're better off using the include to capture the directories that we want and then using the exclude to do more fine-grained filtering.

**Comment by vyasr** [comment]
OK this was bothering me so I went ahead and updated the script to filter live rather than do a batch process at the end, which should make parsing the CI output much easier.

File: cpp/scripts/parse_iwyu_output.py
Comment (davidwendt) [inline cpp/scripts/parse_iwyu_output.py:130]:
```suggestion
```

Comment (vyasr) [reply cpp/scripts/parse_iwyu_output.py:130]: IWYU will error if it does not find all expected sections, so we have to insert this header even though we want the section to be empty. I'll add a comment.

**Review by davidwendt — APPROVED** [review]

File: ci/cpp_linters.sh
Comment (jameslamb) [inline ci/cpp_linters.sh:30]:
```suggestion
git clone --depth 1 https://github.com/include-what-you-use/include-what-you-use.git
```

You only need the tip of this, right? A `--depth-1` would make this clone a little faster (and maybe less susceptible to network issues, because you'd be pulling a smaller amount of data over the wire).

**Review by jameslamb — APPROVED** [review]
Giving you a `ci-codeowners` / `packaging-codeowners` approval. Left one very very tiny suggestion, do whatever you want with it.

File: ci/cpp_linters.sh
Comment (vyasr) [reply ci/cpp_linters.sh:30]: Sure we might as well. Normally I'd be more enthusiastic, but in this case hopefully this clone will go away altogether very soon :) so not as big a deal. No harm improving this in the meantime though.

**Review by karthikeyann — APPROVED** [review]
I tried IWYU years ago. But it was not suitable because of too many changes it suggested.
Good approach to "only include removals".  👍 
Including "tests" would be nice because often some includes are not cleaned up after  development/debug.

**Pull request merged**
